From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Krishna Yarlagadda <kyarlagadda@nvidia.com>
Date: Mon, 26 Jun 2023 08:45:11 +0000
Subject: [PATCH 208/240] NVIDIA: SAUCE: tegra: prod: use legacy prod settings

This is an integration squash commit of following commits:

- [INTG: NVIDIA INTERNAL] phy: tegra: xusb: Support prod-settings properties
- [INTG]: drivers: spi: add support for prod framework
- [INTG: NVIDIA INTERNAL]: mmc: host: Add prod framework changes
- i2c: Add prod settings support
- i2c: tegra: Add new prod setting support
- spi: tegra114: Add new prod settings
- sdhci: tegra: Add new prod settings
- [NV INTERNAL] xusb: tegra: Add new prod setting
- sdhci-tegra: host: Avoid access to prod settings on PreSi.
- tegra: prod: use legacy prod settings

Bug 4359070
Bug 4097475
Bug 4099482
Bug 4189448
Bug 4189442
Bug 4165914
Bug 4165933
Bug 4165918
Bug 4165919
Bug 4765671
Bug 4754882

Signed-off-by: Petlozu Pravareshwar <petlozup@nvidia.com>
Change-Id: I835abc91e1cad79f01b7e113fc8c25ea7584b356
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3198380
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>

diff --git a/drivers/spi/spi-tegra114.c b/drivers/spi/spi-tegra114.c
index 5a8c4a2737b86..c48d0f2fbb430 100644
--- a/drivers/spi/spi-tegra114.c
+++ b/drivers/spi/spi-tegra114.c
@@ -23,6 +23,10 @@
 #include <linux/reset.h>
 #include <linux/spi/spi.h>
 
+#if IS_ENABLED(CONFIG_TEGRA_PROD_LEGACY)
+#include <linux/tegra_prod.h>
+#endif
+
 #define SPI_COMMAND1				0x000
 #define SPI_BIT_LENGTH(x)			(((x) & 0x1f) << 0)
 #define SPI_PACKED				(1 << 5)
@@ -64,6 +68,10 @@
 #define SPI_COMMAND2				0x004
 #define SPI_TX_TAP_DELAY(x)			(((x) & 0x3F) << 6)
 #define SPI_RX_TAP_DELAY(x)			(((x) & 0x3F) << 0)
+#define SPI_COMMAND2_RX_CLK_TAP_DELAY_START	0
+#define SPI_COMMAND2_RX_CLK_TAP_DELAY_WIDTH	6
+#define SPI_COMMAND2_TX_CLK_TAP_DELAY_START	6
+#define SPI_COMMAND2_TX_CLK_TAP_DELAY_WIDTH	6
 
 #define SPI_CS_TIMING1				0x008
 #define SPI_SETUP_HOLD(setup, hold)		(((setup) << 4) | (hold))
@@ -147,6 +155,10 @@
 #define DATA_DIR_TX				(1 << 0)
 #define DATA_DIR_RX				(1 << 1)
 
+#define SPI_MISC				0x194
+#define SPI_MISC_CLKEN_OVERRIDE_START		31
+#define SPI_MISC_CLKEN_OVERRIDE_WIDTH		1
+
 #define SPI_DMA_TIMEOUT				(msecs_to_jiffies(1000))
 #define DEFAULT_SPI_DMA_BUF_LEN			(16*1024)
 #define TX_FIFO_EMPTY_COUNT_MAX			SPI_TX_FIFO_EMPTY_COUNT(0x40)
@@ -220,6 +232,10 @@ struct tegra_spi_data {
 	dma_addr_t				tx_dma_phys;
 	struct dma_async_tx_descriptor		*tx_dma_desc;
 	const struct tegra_spi_soc_data		*soc_data;
+
+#if IS_ENABLED(CONFIG_TEGRA_PROD_LEGACY)
+	struct tegra_prod			*prod_list;
+#endif
 };
 
 static int tegra_spi_runtime_suspend(struct device *dev);
@@ -718,6 +734,33 @@ static void tegra_spi_deinit_dma_param(struct tegra_spi_data *tspi,
 	dma_release_channel(dma_chan);
 }
 
+#if IS_ENABLED(CONFIG_TEGRA_PROD_LEGACY)
+static void tegra_spi_write_prod_settings(struct tegra_spi_data *tspi,
+					  const char *prod_name)
+{
+	int err;
+
+	err = tegra_prod_set_by_name(&tspi->base, prod_name, tspi->prod_list);
+	if (err < 0)
+		dev_dbg_once(tspi->dev,
+			     "Prod config not found for SPI: %d\n", err);
+}
+
+static void tegra_spi_set_prod(struct tegra_spi_data *tspi, int cs)
+{
+	char prod_name[15];
+
+	/* Avoid write to register for transfers to last used device */
+	if (tspi->last_used_cs == cs)
+		return;
+
+	sprintf(prod_name, "prod_c_cs%d", cs);
+	tegra_spi_write_prod_settings(tspi, "prod");
+	tegra_spi_write_prod_settings(tspi, prod_name);
+	tspi->last_used_cs = cs;
+}
+#endif
+
 static int tegra_spi_set_hw_cs_timing(struct spi_device *spi)
 {
 	struct tegra_spi_data *tspi = spi_controller_get_devdata(spi->controller);
@@ -845,6 +888,23 @@ static u32 tegra_spi_setup_transfer_one(struct spi_device *spi,
 				command1 &= ~SPI_CS_SW_VAL;
 		}
 
+#if IS_ENABLED(CONFIG_TEGRA_PROD_LEGACY)
+		if (!tspi->prod_list) {
+			if (tspi->last_used_cs != spi_get_chipselect(spi, 0)) {
+				if (cdata && cdata->tx_clk_tap_delay)
+					tx_tap = cdata->tx_clk_tap_delay;
+				if (cdata && cdata->rx_clk_tap_delay)
+					rx_tap = cdata->rx_clk_tap_delay;
+				command2 = SPI_TX_TAP_DELAY(tx_tap) |
+					   SPI_RX_TAP_DELAY(rx_tap);
+				if (command2 != tspi->def_command2_reg)
+					tegra_spi_writel(tspi, command2, SPI_COMMAND2);
+				tspi->last_used_cs = spi_get_chipselect(spi, 0);
+			}
+		} else {
+			tegra_spi_set_prod(tspi, spi_get_chipselect(spi, 0));
+		}
+#else
 		if (tspi->last_used_cs != spi_get_chipselect(spi, 0)) {
 			if (cdata && cdata->tx_clk_tap_delay)
 				tx_tap = cdata->tx_clk_tap_delay;
@@ -856,6 +916,7 @@ static u32 tegra_spi_setup_transfer_one(struct spi_device *spi,
 				tegra_spi_writel(tspi, command2, SPI_COMMAND2);
 			tspi->last_used_cs = spi_get_chipselect(spi, 0);
 		}
+#endif
 
 	} else {
 		command1 = tspi->command1_reg;
@@ -1335,6 +1396,15 @@ static int tegra_spi_probe(struct platform_device *pdev)
 
 	tspi->host = host;
 	tspi->dev = &pdev->dev;
+
+#if IS_ENABLED(CONFIG_TEGRA_PROD_LEGACY)
+	tspi->prod_list = devm_tegra_prod_get(tspi->dev);
+	if (IS_ERR_OR_NULL(tspi->prod_list)) {
+		dev_dbg(&pdev->dev, "Prod settings list not initialized\n");
+		tspi->prod_list = NULL;
+	}
+#endif
+
 	spin_lock_init(&tspi->lock);
 
 	tspi->soc_data = of_device_get_match_data(&pdev->dev);
@@ -1405,6 +1475,11 @@ static int tegra_spi_probe(struct platform_device *pdev)
 	reset_control_assert(tspi->rst);
 	udelay(2);
 	reset_control_deassert(tspi->rst);
+	tspi->last_used_cs = host->num_chipselect + 1;
+#if IS_ENABLED(CONFIG_TEGRA_PROD_LEGACY)
+	/* initialize CS with 0 in probe */
+	tegra_spi_set_prod(tspi, 0);
+#endif
 	tspi->def_command1_reg  = tegra_spi_readl(tspi, SPI_COMMAND1);
 	tspi->def_command1_reg |= SPI_CS_SEL(tspi->def_chip_select);
 	tspi->def_command1_reg  = SPI_M_S;
@@ -1412,7 +1487,6 @@ static int tegra_spi_probe(struct platform_device *pdev)
 	tspi->spi_cs_timing1 = tegra_spi_readl(tspi, SPI_CS_TIMING1);
 	tspi->spi_cs_timing2 = tegra_spi_readl(tspi, SPI_CS_TIMING2);
 	tspi->def_command2_reg = tegra_spi_readl(tspi, SPI_COMMAND2);
-	tspi->last_used_cs = host->num_chipselect + 1;
 	pm_runtime_put(&pdev->dev);
 	ret = request_threaded_irq(tspi->irq, tegra_spi_isr,
 				   tegra_spi_isr_thread, IRQF_ONESHOT,
-- 
2.43.0

